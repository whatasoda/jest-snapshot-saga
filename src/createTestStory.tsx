import React, { ComponentType, Dispatch, useState, useEffect, FunctionComponent } from 'react';
import { render, act } from '@testing-library/react';
import { ROOT_ELEMENT, MONOLITH_SNAPSHOT } from './utils/symbols';
import {
  StoryOptions,
  TestStoryType,
  TestStorySnapshotState,
  SnapshotSectionRecord,
  MonolithSnapshotState,
} from './decls';
import injectRenderCounter from './utils/injectRenderCounter';
import serialize from './serialize';
import pseudoEventTarget from './mock/pseudoEventTarget';

const start = <P extends object>(Component: ComponentType<P>, { Provider, extraMockFunctions }: StoryOptions = {}) => {
  const { Injected, getRenderCount } = injectRenderCounter(Component);
  let using = false;
  let used = false;

  let dispatchUnmount = (null as unknown) as Dispatch<true>;
  let dispatchProps = (null as unknown) as Dispatch<P>;

  const StoryComponent: FunctionComponent = () => {
    const [props, setProps] = useState<P | null>(null);
    const [unmount, setUnmount] = useState(false);
    useEffect(() => {
      assertUsed('render');
      assertUsing('render');
      using = true;
      dispatchUnmount = setUnmount;
      dispatchProps = setProps;
      return () => {
        used = true;
      };
    }, []);

    if (unmount || !props) return null;
    return <Injected {...props} />;
  };

  const setProps = (props: P) => {
    assertUsed('set props');
    act(() => dispatchProps(props));
  };

  const unmount = () => {
    assertUsed('unmount');
    dispatchUnmount(true);
  };

  const assertUsing = (action: string): void | never => {
    if (using) {
      throw new Error(`Cannot ${action} to a test story being used.`);
    }
  };
  const assertUsed = (action: string): void | never => {
    if (used) {
      throw new Error(`Cannot ${action} to a test story already used.`);
    }
  };

  const result = render(<StoryComponent />, { wrapper: Provider });
  const snapshotState: TestStorySnapshotState = {
    [ROOT_ELEMENT]: true,
    root: result.container,
    isMonolith: false,
    functions: extraMockFunctions ? extraMockFunctions.filter(jest.isMockFunction) : [],
    prev: { elements: '', functions: '', styles: '' },
    diff: { elements: false, functions: false, styles: false },
  };

  return { result, getRenderCount, setProps, unmount, snapshotState };
};

const createTestStory = <P extends object>(
  Component: ComponentType<P>,
  options: StoryOptions = {},
): (() => TestStoryType<P>) => {
  return () => {
    const { getRenderCount, setProps, unmount, snapshotState, result } = start(Component, options);

    const snapshot = expect(snapshotState).toMatchSnapshot;
    const setDiffState = (diff: Partial<SnapshotSectionRecord<boolean>> = {}) => {
      Object.assign(snapshotState.diff, diff);
    };

    return {
      ...pseudoEventTarget,
      setProps,
      unmount,
      snapshot,
      setDiffState,
      getRenderCount,
      result,
    };
  };
};

createTestStory.monolith = <P extends object>(
  Component: ComponentType<P>,
  options: StoryOptions = {},
): (() => TestStoryType<P> & { finish: () => void }) => {
  return () => {
    const { getRenderCount, setProps, unmount, snapshotState, result } = start(Component, options);
    snapshotState.isMonolith = true;

    const monolithSnapshotState: MonolithSnapshotState = {
      [MONOLITH_SNAPSHOT]: true,
      list: [],
    };

    const snapshot = () => {
      monolithSnapshotState.list.push(serialize(snapshotState));
      return snapshotState;
    };
    const setDiffState = (diff: Partial<SnapshotSectionRecord<boolean>> = {}) => {
      Object.assign(snapshotState.diff, diff);
    };

    const finish = expect(monolithSnapshotState).toMatchSnapshot;

    return {
      ...pseudoEventTarget,
      setProps,
      unmount,
      snapshot,
      setDiffState,
      getRenderCount,
      result,
      finish,
    };
  };
};

export default createTestStory;
