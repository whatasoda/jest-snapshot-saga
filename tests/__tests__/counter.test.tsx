import createTestStory from '../../src';

import CounterContainer from '../containers/functional/CounterContainer';
import { act } from '@testing-library/react';

const setDozen = jest.fn().mockName('setDozen');
describe('Counter splitted', () => {
  const start = createTestStory(CounterContainer, {
    extraMockFunctions: [setDozen],
  });

  it('render', () => {
    const story = start();
    story.snapshot();

    story.setProps({ setDozen });
    story.snapshot();
  });
});

describe('Counter', () => {
  const start = createTestStory.monolith(CounterContainer, {
    extraMockFunctions: [setDozen],
  });

  it('renders', async () => {
    const story = start();
    story.snapshot();

    story.setProps({ setDozen });
    story.snapshot();

    const button = await story.result.findByText('Add Item');
    Array.from({ length: 11 }).forEach(() => {
      act(() => button.click());
    });
    story.snapshot();

    act(() => button.click());
    story.snapshot();

    story.finish();
  });
});
