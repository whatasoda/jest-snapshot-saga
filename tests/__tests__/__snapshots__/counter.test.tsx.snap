// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Counter renders 1`] = `
<< 0 <<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<

  >>>>>> functions >>>>>>>>>>>>>>>
    /* none */
  >>>>>> styles >>>>>>>>>>>>>>>>>>
    /* none */
  >>>>>> elements >>>>>>>>>>>>>>>>
    /* none */

<< 1 <<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<

  >>>>>> functions >>>>>>>>>>>>>>>

    setDozen /* #0 */ (
      0
    ) /* return */ => (undefined)

  >>>>>> styles >>>>>>>>>>>>>>>>>>

    .c1 {
      display: -webkit-box;
      display: -webkit-flex;
      display: -ms-flexbox;
      display: flex;
      -webkit-flex-direction: row;
      -ms-flex-direction: row;
      flex-direction: row;
      -webkit-flex-wrap: wrap;
      -ms-flex-wrap: wrap;
      flex-wrap: wrap;
      width: 300px;
    }

    .c0 {
      display: block;
      border-radius: 4px;
      padding: 20px;
    }

  >>>>>> elements >>>>>>>>>>>>>>>>

    <ROOT>
      <a
        class="c0"
      >
        Add Item
      </a>
      <div
        class="c1"
      />
    </ROOT>

<< 2 <<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<

  >>>>>> functions >>>>>>>>>>>>>>>
    /* none */
  >>>>>> styles >>>>>>>>>>>>>>>>>>

    .c1 {
      display: -webkit-box;
      display: -webkit-flex;
      display: -ms-flexbox;
      display: flex;
      -webkit-flex-direction: row;
      -ms-flex-direction: row;
      flex-direction: row;
      -webkit-flex-wrap: wrap;
      -ms-flex-wrap: wrap;
      flex-wrap: wrap;
      width: 300px;
    }

    .c2 {
      width: 15px;
      height: 15px;
      margin: 5px;
    }

    .c0 {
      display: block;
      border-radius: 4px;
      padding: 20px;
    }

  >>>>>> elements >>>>>>>>>>>>>>>>

    <ROOT>
      <a
        class="c0"
      >
        Add Item
      </a>
      <div
        class="c1"
      >
        <div
          class="c2"
        >
          0
        </div>
        <div
          class="c2"
        >
          1
        </div>
        <div
          class="c2"
        >
          2
        </div>
        <div
          class="c2"
        >
          3
        </div>
        <div
          class="c2"
        >
          4
        </div>
        <div
          class="c2"
        >
          5
        </div>
        <div
          class="c2"
        >
          6
        </div>
        <div
          class="c2"
        >
          7
        </div>
        <div
          class="c2"
        >
          8
        </div>
        <div
          class="c2"
        >
          9
        </div>
        <div
          class="c2"
        >
          10
        </div>
      </div>
    </ROOT>

<< 3 <<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<

  >>>>>> functions >>>>>>>>>>>>>>>

    setDozen /* #0 */ (
      1
    ) /* return */ => (undefined)

  >>>>>> styles >>>>>>>>>>>>>>>>>>

    .c1 {
      display: -webkit-box;
      display: -webkit-flex;
      display: -ms-flexbox;
      display: flex;
      -webkit-flex-direction: row;
      -ms-flex-direction: row;
      flex-direction: row;
      -webkit-flex-wrap: wrap;
      -ms-flex-wrap: wrap;
      flex-wrap: wrap;
      width: 300px;
    }

    .c2 {
      width: 15px;
      height: 15px;
      margin: 5px;
    }

    .c0 {
      display: block;
      border-radius: 4px;
      padding: 20px;
    }

  >>>>>> elements >>>>>>>>>>>>>>>>

    <ROOT>
      <a
        class="c0"
      >
        Add Item
      </a>
      <div
        class="c1"
      >
        <div
          class="c2"
        >
          0
        </div>
        <div
          class="c2"
        >
          1
        </div>
        <div
          class="c2"
        >
          2
        </div>
        <div
          class="c2"
        >
          3
        </div>
        <div
          class="c2"
        >
          4
        </div>
        <div
          class="c2"
        >
          5
        </div>
        <div
          class="c2"
        >
          6
        </div>
        <div
          class="c2"
        >
          7
        </div>
        <div
          class="c2"
        >
          8
        </div>
        <div
          class="c2"
        >
          9
        </div>
        <div
          class="c2"
        >
          10
        </div>
        <div
          class="c2"
        >
          11
        </div>
      </div>
    </ROOT>

<< END <<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<
`;

exports[`Counter splitted render 1`] = `
  >>>>>> functions >>>>>>>>>>>>>>>
    /* none */
  >>>>>> styles >>>>>>>>>>>>>>>>>>
    /* none */
  >>>>>> elements >>>>>>>>>>>>>>>>
    /* none */
`;

exports[`Counter splitted render 2`] = `
  >>>>>> functions >>>>>>>>>>>>>>>

    setDozen /* #0 */ (
      0
    ) /* return */ => (undefined)

  >>>>>> styles >>>>>>>>>>>>>>>>>>

    .c1 {
      display: -webkit-box;
      display: -webkit-flex;
      display: -ms-flexbox;
      display: flex;
      -webkit-flex-direction: row;
      -ms-flex-direction: row;
      flex-direction: row;
      -webkit-flex-wrap: wrap;
      -ms-flex-wrap: wrap;
      flex-wrap: wrap;
      width: 300px;
    }

    .c0 {
      display: block;
      border-radius: 4px;
      padding: 20px;
    }

  >>>>>> elements >>>>>>>>>>>>>>>>

    <ROOT>
      <a
        class="c0"
      >
        Add Item
      </a>
      <div
        class="c1"
      />
    </ROOT>
`;
